21.md

```diff
@@ -0,0 +1,96 @@
+        }
+    }
+    ```
+
+    If `SERVICE` supports `LUD-21` and cannot perform the exchange, then an error must be returned.
+    If `SERVICE` returns a payload without `targetExchangeData`, it must be because it doesn't support `LUD-21`.
+
+3. `LN WALLET` displays a payment dialog where user can specify an amount in the fiat currency supported by the wallet:
+
+	```
+	max can send = min(maxSendable, local estimation of how much sats the given amount is)
+	min can send = max(minSendable, local minimal value allowed by wallet)
+	```
+
+    Additionaly, the wallet is able to show how much, approximately, target currency the given amount would be exchanged into.
+    The amount of target fiat currency deposited would be calculated approximately as follows:
+        `sourceAmount / sourcePrice * (1 - sourceFee) * priceHint * (1 - feeHint)`
+
+4. Upon user confirmation, `LN WALLET` buys the sats and makes a GET request using
+
+	```
+	<callback><?|&>amount=<milliSatoshi><?|&>sourceExchangeData=<urlencode({json object})>
+	```
+
+    `amount` being the amount of milisats to be sent after successful conversion.
+
+    `sourceExchangeData` being an optional parameter with the following format
+
+    ```Typescript
+    {
+        "currency": string, // The currency the sats were bought with.
+        "price": number, // The exact price paid to buy the BTC in the source currency.
+        "fee": number | null // The exact percentage fee the `LN WALLET` has charged to convert BTC to currency.
+    }
+    ```
+
+    If omitted, the `SERVICE` performs the conversion without knowing the source currency.
+    If provided, the `SERVICE` is able to provide richer information about the conversion flow.
+
+5. `LN Service` takes the GET request and returns JSON response of form:
+
+	```Typescript
+	{
+	    pr: string, // bech32-serialized lightning invoice
+        routes: [] // an empty array
+	}
+	```
+
+6. `LN WALLET` Verifies that `h` tag in provided invoice is a hash of `metadata` string converted to byte array in UTF-8 encoding.
+7. `LN WALLET` Verifies that amount in provided invoice equals the amount previously specified by user.
+8. `LN WALLET` pays the invoice, no additional user confirmation is required at this point.
+9. `LN SERVICE` receives the payment and automatically converts the BTC into the currency `LN WALLET` specified when making the `payRequest`, crediting the `LN SERVICE` user in fiat, not in BTC.
+
+## Notes on price hints and better user experience
+
+The interaction described in this document never provides any sort of guarantees about price. That said, the closer the `SERVICE` and `WALLET` can get the hints to the actual exchange price, the better the end user experience. This means `SERVICE` should try to give as precise as possible price hints while `WALLET` should refresh the hints as needed during the user transfer flow.
```

Comment:
As discussed offline: if there are no guarantees about the price, it could be worth for `LN SERVICE` and `LN WALLET` to report to each other the exact price used after the transaction completes.

Reply:
@joostjager Updated the spec, can you tell me what you think?